File: src/temp/ast/expression/Literal.java
package temp.ast.expression;

import temp.ast.Expression;
import temp.primitives.PrimitiveType;

public class Literal extends Expression {
    private final String value;
    private final PrimitiveType type;

    private final int line;

    public Literal(String value, PrimitiveType type, int line) {
        this.value = value;
        this.type = type;
        this.line = line;
    }

    public <R> R accept(Visitor<R> visitor) {
        return visitor.visitLiteralExpression(this);
    }

    public String getValue() {
        if (this.value == null) {
            return "null";
        }

        return this.value;
    }

    public int getLine() {
        return this.line;
    }

    public PrimitiveType getType() {
        return this.type;
    }
}

File: src/codepulsar/ast/expression/Literal.cpp
#include "Literal.h"


Pulsar::Literal::Literal(std::string value, PrimitiveType type, int line) {
    this->value = value;
    this->type = type;
    this->line = line;
}

std::any Pulsar::Literal::accept(ExprVisitor& visitor) {
    return visitor.visitLiteralExpression(this);
}

std::string Pulsar::Literal::getValue() {
    return this->value;
}

Pulsar::PrimitiveType Pulsar::Literal::getType() {
    return this->type;
}

int Pulsar::Literal::getLine() {
    return this->line;
}
